soci/soci_index.go

```diff
@@ -333,24 +333,46 @@ func (b *IndexBuilder) Build(ctx context.Context, img images.Image) (*IndexWithM
 		return nil, err
 	}
 
-	sociLayersDesc := make([]*ocispec.Descriptor, len(manifest.Layers))
-	eg, ctx := errgroup.WithContext(ctx)
-	for i, l := range manifest.Layers {
-		i, l := i, l
-		eg.Go(func() error {
+	sociLayersDesc := make([]*ocispec.Descriptor, 0, len(manifest.Layers))
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(manifest.Layers))
```

Comment:
This change (336-368) adds: (1) a mutex (extra overhead); (2) a waitgroup (manually add/done); (3) error channel (extra complication).

Personally I still prefer the existing `errgroup` approach which is simpler by using index instead of mutex + append when saving `sociLayersDesc`.

I suggest add a smilar `errs` slice (`make(error, 0, len(manifest.Layers))`) that stores all errors, and you can for-loop and print non-nil errors after `eg.Wait()`. This can avoid all the three additions above.

Reply:
I think that `errgroup` can only produce the first error emitted by any goroutine in the group. We need to collect all of the errors. At the moment we're just printing them, which could happen in the goroutine and not get returned and avoid that need, but I anticipate we will actually collect them for error aggregation soon.

Reply:
I would find it misleading to use an errgroup that doesn't actually capture errors.
The latest code eliminates the mutex, but still has a bit more code than the errgroup.

Reply:
How about something like the following?

Pros
1. smaller channel buffer
2. start printing errors as soon as they are sent thru.

```
ch := make(chan error, 1)

go func() {
  var wg sync.WaitGroup
  for ... {
    wg.Add(1)
    go func() {
      defer wg.Done()
      ...
      if err != nil {
        ch <- err
      }
    }()
  }
  wg.Wait()
  close(ch)
}()

for err := range ch {
  fmt.Println(err)
}
```

Reply:
I don't think that would start printing errors immediately. the wg.Wait() would block until the whole group was finished, but if two things produce errors then their channel write would block because of the buffer size.

Reply:
wg.Wait() (and channel close) are in a separate goroutine. The main goroutine will just go from `ch := make(..)` to the for loop over error channel until the channel is closed (after wg.Wait())

Reply:
I have implemented most of Henry's suggestion, although the error accumulation is still difference per Kern's feedback below that we should be returning an error. I decided to wrap all the errors, rather than just print them and then return a single generic error.